### PR TITLE
Linux test fixes

### PR DESCRIPTION
**Pull request by slide**
More linux test fixes (and some disabled)

### CODE REVIEW

File: Solutions/Common.proj [hunk 1]
```diff
@@ -57,7 +57,8 @@
...
+    <Features Condition="'$(Mono)' == 'true'">FEATURE_CUSTOM_MODIFIERS;FEATURE_NATIVE;FEATURE_EXCEPTION_STATE;FEATURE_DEFAULT_PARAMETER_VALUE;FEATURE_CRYPTOGRAPHY;FEATURE_SYNC_SOCKETS;FEATURE_ASSEMBLY_RESOLVE;FEATURE_CUSTOM_TYPE_DESCRIPTOR;FEATURE_ENCODING;FEATURE_DBNULL;FEATURE_FILESYSTEM;FEATURE_THREAD;FEATURE_PROCESS;FEATURE_PIPES;FEATURE_STACK_TRACE;FEATURE_TYPECONVERTER;FEATURE_PDBEMIT;FEATURE_COM;FEATURE_FULL_CONSOLE;FEATURE_BASIC_CONSOLE;FEATURE_SERIALIZATION;FEATURE_REMOTING;FEATURE_CONFIGURATION;FEATURE_REFEMIT;FEATURE_CODEDOM;FEATURE_COMPRESSION;FEATURE_WARNING_EXCEPTION;FEATURE_APTCA;FEATURE_WIN32EXCEPTION;FEATURE_ASSEMBLY_LOCATION;FEATURE_ENUM_NAMES_VALUES;FEATURE_OS_SERVICEPACK;FEATURE_APPLICATIONEXCEPTION;FEATURE_STOPWATCH;FEATURE_APARTMENTSTATE;FEATURE_GETINVALIDFILENAMECHARS;FEATURE_LOADWITHPARTIALNAME;FEATURE_DRIVENOTFOUNDEXCEPTION;FEATURE_ASSEMBLY_CODEBASE;FEATURE_FULL_NET;FEATURE_FULL_CRYPTO;FEATURE_ICLONEABLE;FEATURE_ANSICP;FEATURE_TIMEZONE;FEATURE_SORTKEY;FEATURE_UNIX</Features>
```
Comment (slozier) [inline Solutions/Common.proj]: Any reason why `FEATURE_COM` is excluded in 4.0 and 4.5 but not here?

File: Solutions/Common.proj [hunk 2]
```diff
@@ -57,7 +57,8 @@
   <!-- References -->
   <PropertyGroup Condition="'$(ReferencedPlatform)' == 'V2'">
     <TargetFrameworkVersion>v3.5</TargetFrameworkVersion>
-    <Features>FEATURE_CUSTOM_MODIFIERS;FEATURE_NATIVE;FEATURE_EXCEPTION_STATE;FEATURE_DEFAULT_PARAMETER_VALUE;FEATURE_CRYPTOGRAPHY;FEATURE_SYNC_SOCKETS;FEATURE_ASSEMBLY_RESOLVE;FEATURE_CUSTOM_TYPE_DESCRIPTOR;FEATURE_ENCODING;FEATURE_DBNULL;FEATURE_FILESYSTEM;FEATURE_THREAD;FEATURE_PROCESS;FEATURE_PIPES;FEATURE_STACK_TRACE;FEATURE_TYPECONVERTER;FEATURE_PDBEMIT;FEATURE_COM;FEATURE_FULL_CONSOLE;FEATURE_BASIC_CONSOLE;FEATURE_SERIALIZATION;FEATURE_REMOTING;FEATURE_CONFIGURATION;FEATURE_REFEMIT;FEATURE_CODEDOM;FEATURE_COMPRESSION;FEATURE_WARNING_EXCEPTION;FEATURE_APTCA;FEATURE_WIN32EXCEPTION;FEATURE_ASSEMBLY_LOCATION;FEATURE_ENUM_NAMES_VALUES;FEATURE_OS_SERVICEPACK;FEATURE_APPLICATIONEXCEPTION;FEATURE_STOPWATCH;FEATURE_APARTMENTSTATE;FEATURE_GETINVALIDFILENAMECHARS;FEATURE_LOADWITHPARTIALNAME;FEATURE_DRIVENOTFOUNDEXCEPTION;FEATURE_ASSEMBLY_CODEBASE;FEATURE_FULL_NET;FEATURE_FULL_CRYPTO;FEATURE_ICLONEABLE;FEATURE_ANSICP;FEATURE_TIMEZONE;FEATURE_SORTKEY</Features>
```
Comment (slozier) [inline Solutions/Common.proj:60]: Can we do this differently? Perhaps something like (assuming it works, I didn't actually test):
```
<Features>...</Features>
<Features Condition="'$(Mono)' != 'true'">$(Features);FEATURE_COM;FEATURE_WINDOWS</Features>
<Features Condition="'$(Mono)' == 'true'">$(Features);FEATURE_UNIX</Features>
```

File: Solutions/Common.proj [hunk 3]
```diff
@@ -57,7 +57,8 @@
...
+    <Features Condition="'$(Mono)' != 'true'">FEATURE_CUSTOM_MODIFIERS;FEATURE_NATIVE;FEATURE_EXCEPTION_STATE;FEATURE_DEFAULT_PARAMETER_VALUE;FEATURE_CRYPTOGRAPHY;FEATURE_SYNC_SOCKETS;FEATURE_ASSEMBLY_RESOLVE;FEATURE_CUSTOM_TYPE_DESCRIPTOR;FEATURE_ENCODING;FEATURE_DBNULL;FEATURE_FILESYSTEM;FEATURE_THREAD;FEATURE_PROCESS;FEATURE_PIPES;FEATURE_STACK_TRACE;FEATURE_TYPECONVERTER;FEATURE_PDBEMIT;FEATURE_COM;FEATURE_FULL_CONSOLE;FEATURE_BASIC_CONSOLE;FEATURE_SERIALIZATION;FEATURE_REMOTING;FEATURE_CONFIGURATION;FEATURE_REFEMIT;FEATURE_CODEDOM;FEATURE_COMPRESSION;FEATURE_WARNING_EXCEPTION;FEATURE_APTCA;FEATURE_WIN32EXCEPTION;FEATURE_ASSEMBLY_LOCATION;FEATURE_ENUM_NAMES_VALUES;FEATURE_OS_SERVICEPACK;FEATURE_APPLICATIONEXCEPTION;FEATURE_STOPWATCH;FEATURE_APARTMENTSTATE;FEATURE_GETINVALIDFILENAMECHARS;FEATURE_LOADWITHPARTIALNAME;FEATURE_DRIVENOTFOUNDEXCEPTION;FEATURE_ASSEMBLY_CODEBASE;FEATURE_FULL_NET;FEATURE_FULL_CRYPTO;FEATURE_ICLONEABLE;FEATURE_ANSICP;FEATURE_TIMEZONE;FEATURE_SORTKEY;FEATURE_WINDOWS</Features>
```
Comment (slozier) [inline Solutions/Common.proj]: I'm not sure I like the Mono conditional to determine if we're doing a windows/linux build but I guess it's fine for now...

File: External.LCA_RESTRICTED/Languages/IronPython/27/Lib/test/test_mmap.py [hunk 4]
```diff
@@ -543,6 +546,7 @@ def __new__(klass, *args, **kwargs):
         anon_mmap(PAGESIZE)
 
     @unittest.skipUnless(hasattr(mmap, 'PROT_READ'), "needs mmap.PROT_READ")
+    @unittest.skipIf(sys.platform=='cli' and os.name=='posix', 'IronPython does not support this yet')
```
Comment (slozier) [inline External.LCA_RESTRICTED/Languages/IronPython/27/Lib/test/test_mmap.py]: I wonder if we should include issue numbers for these types of exclusion, would make finding these things easier later down the line.

File: Languages/IronPython/IronPythonTest/EngineTest.cs [hunk 5]
```diff
@@ -1031,7 +1031,7 @@ class EmptyNC(object): pass
         public void ScenarioDlrInterop() {
             string actionOfT = typeof(Action<>).FullName.Split('`')[0];
 
-#if !SILVERLIGHT
+#if FEATURE_WINDOWS
```
Comment (slozier) [inline Languages/IronPython/IronPythonTest/EngineTest.cs]: Which part of this requires Windows?

**Review by slozier — COMMENTED** [review]
The only issue I really have with all this is the use of `posix` conditionals to mean `Mono`.

File: Languages/IronPython/IronPythonTest/EngineTest.cs (hunk 5 above)
Comment (slide) [reply Languages/IronPython/IronPythonTest/EngineTest.cs]: The Windows forms stuff

File: Solutions/Common.proj (hunk 1 above)
Comment (slide) [reply Solutions/Common.proj]: No, I don't know why. It was done before my time.

File: External.LCA_RESTRICTED/Languages/IronPython/27/Lib/test/test_mmap.py (hunk 4 above)
Comment (slide) [reply External.LCA_RESTRICTED/Languages/IronPython/27/Lib/test/test_mmap.py]: That's a good idea

File: Solutions/Common.proj (hunk 2 above)
Comment (slide) [reply Solutions/Common.proj:60]: Yes, we do it in ipy3, I'll update.

File: Solutions/Common.proj (hunk 3 above)
Comment (slide) [reply Solutions/Common.proj]: I agree, it's not ideal

File: Languages/IronPython/IronPythonTest/EngineTest.cs (hunk 5 above)
Comment (slozier) [reply Languages/IronPython/IronPythonTest/EngineTest.cs]: I thought Mono supported `System.Windows.Forms`. Perhaps we should create a `FEATURE_WINFORMS` instead?

In any case, only a small part of this test actually requires winforms. I will see if I can refactor it (unless you want to do it).

Comment (slide) [reply Languages/IronPython/IronPythonTest/EngineTest.cs]: It does support WinForms, but not when the build server is headless.

File: Languages/IronPython/Tests/test_file.py [hunk 6]
```diff
@@ -385,94 +385,91 @@ def test_newlines_attribute():
...
+        Assert(f.fileno() <> 0)
 
-    # write
-    AssertError(TypeError, f.writelines, [3])
-    f.writelines(["firstline\n"])
+        # write
+        AssertError(TypeError, f.writelines, [3])
+        f.writelines(["firstline\n"])
 
     f.close()
```
Comment (slozier) [inline Languages/IronPython/Tests/test_file.py]: These should probably be inside the with statement.

File: Util/IronPython/Lib/iptest/console_util.py [hunk 7]
```diff
@@ -105,7 +105,7 @@ def EatToMarker(self, marker, readError=False):
     # Execute a single-line command, and return the output
     def ExecuteLine(self, line, readError=False):
         self.writer.Write(line+"\n")
-        return self.EatToPrompt(readError)[0:-2]
+        return self.EatToPrompt(readError).rstrip()
```
Comment (slozier) [inline Util/IronPython/Lib/iptest/console_util.py:108]: Unrelated to this PR, but it looks like we have some duplication. Any idea why this file appears in two locations (Util\IronPython\Lib\iptest and Languages\IronPython\IronPython\Lib\iptest)?

Comment (slide) [reply Util/IronPython/Lib/iptest/console_util.py:108]: No, I was wondering the same thing. Was going to look into it.

Comment (slozier) [reply Util/IronPython/Lib/iptest/console_util.py:108]: Looks like it may be for the `ClrType_ipy` test (although it's disabled, probably because the script is missing).

File: Languages/IronPython/IronPython/Modules/sys.cs [hunk 8]
```diff
@@ -177,7 +177,7 @@ public static class SysModule {
 
         public static object getfilesystemencoding() {
             if(Environment.OSVersion.Platform == PlatformID.Unix)
-                return "utf-8";
+                return "UTF-8";
```
Comment (slozier) [inline Languages/IronPython/IronPython/Modules/sys.cs]: Why the change to uppercase? In https://docs.python.org/2/library/sys.html#sys.getfilesystemencoding it appears to be lowercase.

Comment (slide) [reply Languages/IronPython/IronPython/Modules/sys.cs]: In practice it is not lowercase. It's uppercase in cpython on my linux box

Comment (slozier) [reply Languages/IronPython/IronPython/Modules/sys.cs]: Is there a test that checks specifically for this? In my Ubuntu docker container it returns `ANSI_X3.4-1968`.

Comment (slide) [reply Languages/IronPython/IronPython/Modules/sys.cs]: No, I'll revert, it's not necessary.
